tools/webtransport/h3/webtransport_h3_server.py

```diff
@@ -155,24 +216,46 @@ def stash(self) -> stash.Stash:
             self._stash = stash.Stash(self._stash_path, address, authkey)
         return self._stash
 
+    @property
+    def dict_for_handlers(self) -> Dict[str, Any]:
+        """A dictionary that handlers can attach arbitrary data."""
+        return self._dict_for_handlers
+
     def stream_is_unidirectional(self, stream_id: int) -> bool:
         """Return True if the stream is unidirectional."""
         return stream_is_unidirectional(stream_id)
 
-    def close(self,
-              error_code: int = QuicErrorCode.NO_ERROR,
-              reason_phrase: str = "") -> None:
+    def close(self, close_info: Optional[Tuple[int, bytes]]) -> None:
         """
         Close the session.
 
-        :param error_code: An error code indicating why the session is
-                           being closed.
-        :param reason_phrase: A human readable explanation of why the
-                              session is being closed.
+        :param close_info The close information to send.
         """
-        self._http._quic.close(error_code=error_code,
-                               reason_phrase=reason_phrase)
+        self._allow_calling_session_closed = False
+        assert self._session_stream_id is not None
+        if close_info is not None:
+            code = close_info.code
+            reason = close_info.reason
+            buffer = Buffer(capacity=len(reason) + 4)
+            buffer.push_uint32(code)
+            buffer.push_bytes(reason)
+            capsule = H3Capsule(CLOSE_WEBTRANSPORT_SESSION, buffer.data)
+            self.send_stream_data(self._session_stream_id, capsule.encode())
+
+        self.send_stream_data(self._session_stream_id, b'', end_stream=True)
         self._protocol.transmit()
+        # TODO(yutakahirano): Reset all other streams.
+        # TODO(yutakahirano): Reject future stream open requests
+        # We need to wait for the stream data to arrive at the client, and then
+        # we need to close the connection. At this moment we're relying on the
+        # client's behavior.
+        # TODO(yutakahirano): Implement the above.
+
+    def call_session_closed(self, close_info: Optional[Tuple[int, bytes]], abruptly: bool) -> None:
+        allow_calling_session_closed = self._allow_calling_session_closed
```

Comment:
My mental model is that WebTransportSession has a state machine (e.g. not established, established, closed) and decides behavior based on the state so I slightly prefer to have an enum or `_session_closed` flag, but this approach will also work.